db/sqlinterfaces.c

```diff
@@ -1795,9 +1793,13 @@ int handle_sql_begin(struct sqlthdstate *thd, struct sqlclntstate *clnt,
                 (clnt->sql) ? clnt->sql : "(???.)");
 
     /* Latch the last commit LSN */
-    if (clnt->dbtran.mode == TRANLEVEL_MODSNAP && (populate_modsnap_state(clnt) != 0)) {
-        rc = SQLITE_INTERNAL;
-        goto done;
+    if (clnt->dbtran.mode == TRANLEVEL_MODSNAP) {
+        rdlock_schema_lk();
+        rc = populate_modsnap_state(clnt) ? SQLITE_INTERNAL : 0;
+        unlock_schema_lk();
+        if (rc) {
+            goto done;
+        }
```

Comment:
there should be no table locks held here. the client has only just ran `begin`, so we don't know what tables they are going to access.